tools-local/Microsoft.ML.InternalCodeAnalyzer/BestFriendOnPublicDeclarationsAnalyzer.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.ML.InternalCodeAnalyzer
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class BestFriendOnPublicDeclarationsAnalyzer : DiagnosticAnalyzer
+    {
+        private const string Category = "Access";
+        internal const string DiagnosticId = "MSML_BestFriendOnPublicDeclaration";
+
+        private const string Title = "Public declarations should not have " + AttributeName + " attribute.";
+        private const string Format = "The " + AttributeName + " should not be applied to publicly visible members.";
+
+        private const string Description =
+            "The " + AttributeName + " attribute is not valid on public identifiers.";
+
+        private static DiagnosticDescriptor Rule =
+            new DiagnosticDescriptor(DiagnosticId, Title, Format, Category,
+                DiagnosticSeverity.Warning, isEnabledByDefault: true, description: Description);
+
+        private const string AttributeName = "Microsoft.ML.BestFriendAttribute";
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(CompilationStart);
+        }
+
+        private void CompilationStart(CompilationStartAnalysisContext context)
+        {
+            var list = new List<string> { AttributeName, "Microsoft.ML.Internal.CpuMath.Core.BestFriendAttribute" };
+
+            foreach (var attributeName in list)
+            {
+                var attribute = context.Compilation.GetTypeByMetadataName(attributeName);
+
+                if (attribute == null)
+                    continue;
+                
+                context.RegisterSymbolAction(c => AnalyzeCore(c, attribute), SymbolKind.NamedType, SymbolKind.Method, SymbolKind.Field, SymbolKind.Property);
+            }
+        }
+
+        private void AnalyzeCore(SymbolAnalysisContext context, INamedTypeSymbol attributeType)
+        {
+            if (context.Symbol.DeclaredAccessibility != Accessibility.Public)
```

Comment:
📝 This will report a diagnostic on something like the following. You'll want to review and determine if this is desirable.

```csharp
internal class NotPublicType
{
  [BestFriend]
  public void DeclaredPublicButNotExposed() { }
}
```

Reply:
That is a good point. The effective visibility is `internal` here but a declaration like this is still not "correct". How do you imagine we handle this, @TomFinley?

We could either just keep reporting this as incorrect (simple 😄) or recursively compute the effective visibility. That would require us to walk all the way to the assembly level for each symbol. That might get expensive.

Reply:
Hi just saw this. If it will report something like this, that is good, not bad. So I'd say @mareklinka's change here is desirable, if it in fact does warn on this.

Indeed, it's unclear to me what the above snippet would do? My best guess is, the thinking that references to `NotPublicType` should fail, but references to `NotPublicType.DeclaredPublicButNotExposed` should succeed? If my guess as to what the point is is correct, that is against the spirit of #1519 which is about making it obvious when one of these "cross-assembly-infrastructure" types and members were being declared. So the snippet's usage of `[BestFriend]` is functionally not useful (since it serves no purpose), which means that I'd think it should warn.